[TRTLLM-6091][docs] Update docs/trtllm sampler 1.0

stnie opened the pull request:
## Description

Added Documentation for TRTLLM Sampling. Reused the old Documentation and added a new section regarding Beam Search

## Test Coverage

<!--
Please list clearly what are the relevant test(s) that can safeguard the changes in the PR. This helps us to ensure we have sufficient test coverage for the PR.
-->

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--disable-fail-fast --skip-test --stage-list "A10-1, xxx" --gpu-type "A30, H100_PCIe" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-[Post-Merge]-1, xxx"]`

Launch build/test pipelines. All previously running jobs will be killed.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests. Will also run L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-[Post-Merge]-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-[Post-Merge]-1, xxx".

For guidance on mapping tests to stage names, see `docs/source/reference/ci-overview.md`.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

stnie committed:
docs: added beam search feature documentation for TRTLLM Sampler and updated sampler documentation

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

chenopis commented on docs/source/1.0/features/sampling.md:
```diff
@@ -0,0 +1,54 @@
+Parameter Configuration:
+- `best_of`: Controls the number of beams processed during generation (beam width)
+- `n`: Controls the number of output sequences returned (can be less than `best_of`)
+- If `best_of` is omitted, it will be implicitly set to `n`
```
Phrase it as a default, i.e.

- If `best_of` is omitted, the number of beams processed defaults to `n`

stnie replied:
I updated the test passage

chenopis reviewed (approved):
Some small suggested edits.

stnie committed:
docs: update sampling documentation for clarity and detail

- Clarified usage instructions for enabling the `enable_trtllm_sampler` option and passing `SamplingParams`.
- Improved wording for better readability, including changes to log probability terminology.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie commented on docs/source/1.0/features/sampling.md:
```diff
@@ -0,0 +1,54 @@
+# Sampling
+The PyTorch backend supports most of the sampling features that are supported on the C++ backend, such as temperature, top-k and top-p sampling, beam search, stop words, bad words, penalty, context and generation logits, and log probs.
```
I updated the test passage

stnie commented on docs/source/1.0/features/sampling.md:
```diff
@@ -0,0 +1,54 @@
+
+## General usage
+
+In order to use this feature, it is necessary to enable option `enable_trtllm_sampler` in the `LLM` class, and pass a `SamplingParams` object with the desired options as well. The following example prepares two identical prompts which will give different results due to the sampling parameters chosen:
```
I updated the test passage

stnie commented on docs/source/1.0/features/sampling.md:
````diff
@@ -0,0 +1,54 @@
+            "Hello, my name is"], sampling_params)
+```
+
+When using speculative decoders such as MTP or Eagle-3, the `enable_trtllm_sampler` option is not yet supported and therefore the subset of sampling options available is more restricted.
````
I updated the test passage

stnie committed:
doc: Update Beam Search documentation

Added instructions regarding CUDA Graphs to the documentation of sampling with beam search.

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

chenopis reviewed (approved):
LGTM

chenopis reviewed (approved):
LGTM

syuoni reviewed (changes requested):
Hi @stnie ,

Sorry, I revisited the [doc plan](https://docs.google.com/spreadsheets/d/1PfY5M87TanVUf9hbUwkElQFr7M8J_kw_WqMoYNumiIU/edit?gid=0#gid=0) 33rd row: Sampling. It should cover greedy, beam, logit processor, guided decoding, xgrammar, etc.

Seems the logits processor, guided decoding (i.e., xgrammar/llguidance) are missing from the PR. Could you please add them? Thanks

stnie committed:
doc: Added documentation for guided decoding and Logits Processor

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie committed:
doc: updated guided decoding and logits processing sections of the documentation

-Removed sections covering BatchedLogitsProcessor
- updated signature of __call__ function in the LogitsProcessor example
- added cross references to specific classes mentioned in the documentation
- Updated Guided decoding section, to differ between JSON schema and JSON object and mentioned llguidance backend support

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie committed:
doc: updated sampling documentation

- removed disable_overlap_scheduler from guided config
- added additional imports to logits processor example

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>

stnie commented on docs/source/1.0/features/sampling.md:
```diff
@@ -17,4 +23,138 @@ llm.generate(["Hello, my name is",
+Guided decoding controls the generation outputs to conform to pre-defined structured formats, ensuring outputs follow specific schemas or patterns.
+
+The PyTorch backend supports guided decoding with the [XGrammar](https://github.com/mlc-ai/xgrammar) backend and the following formats:
+- JSON
```
I added both options to the list

stnie commented on docs/source/1.0/features/sampling.md:
```diff
@@ -17,4 +23,138 @@ llm.generate(["Hello, my name is",
+
+To enable guided decoding, you must:
+
+1. Set the `guided_decoding_backend` parameter to `'xgrammar'` in the `LLM` class
```
I added it onto the general description as well as the step-by-step list

stnie commented on docs/source/1.0/features/sampling.md:
```diff
@@ -17,4 +23,138 @@ llm.generate(["Hello, my name is",
+
+The PyTorch backend supports two types of logits processing:
+- **Per-prompt processing**: Processes logits for each prompt individually, providing fine-grained control
+- **Batched processing**: Processes logits for the entire batch together, optimizing performance for larger batch sizes
```
removed

stnie commented on docs/source/1.0/features/sampling.md:
````diff
@@ -17,4 +23,138 @@ llm.generate(["Hello, my name is",
+llm.generate(["Hello, my name is"], sampling_params)
+```
+
+### Batched logits processing
````
removed

stnie commented on docs/source/1.0/features/sampling.md:
```diff
@@ -17,4 +23,138 @@ llm.generate(["Hello, my name is",
+To enable guided decoding, you must:
+
+1. Set the `guided_decoding_backend` parameter to `'xgrammar'` in the `LLM` class
+2. Disable overlap scheduling using the `disable_overlap_scheduler` parameter of the `LLM` class
```
adjusted the examples and the step-by-step list to no longer disable overlap scheduler

syuoni commented on docs/source/1.0/features/sampling.md:
```diff
@@ -17,4 +23,113 @@ llm.generate(["Hello, my name is",
+        client_id: Optional[int]
+    ) -> None:
+        # Implement your custom inplace logits processing logic
+        logits = 2 * logits
```
Must use in-place op so that the logits processing is effective.

```suggestion
        logits *= logits
```

stnie replied:
done

syuoni commented on docs/source/1.0/features/sampling.md:
```diff
@@ -17,4 +23,113 @@ llm.generate(["Hello, my name is",
+    * Note: Depending on the type of format, a different parameter needs to be chosen to construct the object (`json`, `regex`, `grammar`, `strucutral_tag`).
+3. Pass the `GuidedDecodingParams` object to the `guided_decoding` parameter of the `SamplingParams` object
+
+The following example demonstrates guided decoding with JSON format:
```
```suggestion
The following example demonstrates guided decoding with a JSON schema:
```

stnie replied:
done

syuoni commented on docs/source/1.0/features/sampling.md:
```diff
@@ -17,4 +23,113 @@ llm.generate(["Hello, my name is",
+sampling_params = SamplingParams(
+        guided_decoding=guided_decoding_params,
+    )
+llm.generate(["Generate a JSON response"], sampling_params)
```
```suggestion
llm.generate("Generate a JSON response", sampling_params)
```

syuoni reviewed (commented):
LGTM, thanks!

stnie committed:
chore: include requested changes to sampling documentation

- fixed LogitsProcessor example to actually modify logits inplace
- small text adjustments

Signed-off-by: Stefan Niebler <82932102+stnie@users.noreply.github.com>
